### DISCUSSION TITLE
Multiple classes in ClassConfig

### DISCUSSION THREAD

**Discussion by natedailey7** [azavea/raster-vision#2027] [Q&A] [answered]
I'm trying to input 2 classes into my Raster Vision model, but I can't figure out how to pass in my class ID from my geojson. For example, my geojson file class ID is named `'id'`, with the classes 0 and 1.
```
class_config = ClassConfig(
    names=['0', '1'],
    colors=['lightgreen', 'darkred'],
    )
```
Is there a place I can specify the field name `'id'`, so that Raster Vision knows where to look for the class ID?

**Comment by AdeelH** [comment]
You will need to use the [`ClassInferenceTransformer`](https://docs.rastervision.io/en/stable/api_reference/_generated/rastervision.core.data.vector_transformer.class_inference_transformer.ClassInferenceTransformer.html#rastervision.core.data.vector_transformer.class_inference_transformer.ClassInferenceTransformer) and pass it a `class_id_to_filter` ([example usage](https://docs.rastervision.io/en/stable/usage/tutorials/stac_plus_osm.html#Read-in-the-OSM-labels-as-a-LabelSource)). In your case, the `class_id_to_filter` should be something like `class_id_to_filter={0: ['==', 'id', 0], 1: ['==', 'id', 1]}`. In case it is unclear where to pass the transformer, please share more of your code and I can help you with that.

**Reply by natedailey7** [reply]
Thanks for your response! That makes sense, adding `class_id_to_filter` to a `GeoJSONVectorSource`. I'm not sure how to apply this to the `ObjectDetectionSlidingWindowGeoDataset.from_uris()` function. 

That is, I want to add `class_id_to_filter={0: ['==', 'id', 0], 1: ['==', 'id', 1]}` to the following code:
```python
class_config = ClassConfig(
    names=['0', '1'],
    colors=['lightgreen', 'darkred'],
    )

train_image_uri = "path"
train_label_uri = "path"
val_image_uri   = "path"
val_label_uri   = "path"

data_augmentation_transform = A.Compose([
    A.Flip(p=0.5),
    A.HorizontalFlip(p=0.5),
    ], bbox_params=DEFAULT_BBOX_PARAMS
)

train_ds = ObjectDetectionSlidingWindowGeoDataset.from_uris(
    class_config=class_config,
    image_uri=train_image_uri,
    label_vector_uri=train_label_uri,
    transform=data_augmentation_transform,
    size=400,
    stride=400,
)
```
I notice the parameter `transform` for `ObjectDetectionSlidingWindowGeoDataset.from_uris()`, however it seems that this is just for data augmentation transforms?

**Reply by AdeelH** [reply]
You need to pass it to the `label_source_kw` argument (see [docs](https://docs.rastervision.io/en/0.21/api_reference/_generated/rastervision.pytorch_learner.dataset.object_detection_dataset.ObjectDetectionSlidingWindowGeoDataset.html#rastervision.pytorch_learner.dataset.object_detection_dataset.ObjectDetectionSlidingWindowGeoDataset.from_uris)):

```py
train_ds = ObjectDetectionSlidingWindowGeoDataset.from_uris(
    class_config=class_config,
    image_uri=train_image_uri,
    label_vector_uri=train_label_uri,
    label_vector_source_kw=dict(
        transformers=ClassInferenceTransformerConfig(
            class_id_to_filter={0: ['==', 'id', 0], 1: ['==', 'id', 1]})),
    transform=data_augmentation_transform,
    size=400,
    stride=400,
)
```

`ClassInferenceTransformerConfig` also accepts a `default_class_id` which is the catch-all ID for features in the GeoJSON whose class ID cannot be inferred by other means. If you set it to `None` (the default) such features will be ignored.

**Reply by natedailey7** [reply]
I'm getting the following error. Do you have any recommendations?
```
ValidationError: 1 validation error for GeoJSONVectorSourceConfig
transformers
  value is not a valid list (type=type_error.list)
```
This is after adding
```python
    label_vector_source_kw=dict(
        transformers=ClassInferenceTransformerConfig(
            class_id_to_filter={0: ['==', 'id', 0], 1: ['==', 'id', 1]})),
```
To my `ObjectDetectionSlidingWindowGeoDataset.from_uris()`

**Reply by AdeelH** [reply, marked as answer]
Oops, the value passed to `transformers` needs to be a list:
```py
    label_vector_source_kw=dict(
        transformers=[ClassInferenceTransformerConfig(
            class_id_to_filter={0: ['==', 'id', 0], 1: ['==', 'id', 1]}))],
```

**Reply by natedailey7** [reply]
That worked! Thank you so much! 🙇
